v2/tools/generator/internal/conversions/README.md

```diff
@@ -0,0 +1,568 @@
+
+``` go
+sourceEndpoint := &TypedConversionEndpoint {
+    name:    "Quantity",
+    theType: IntType,
+}
+
+destinationEndpoint := &TypedConversionEndpoint {
+    name:    "Quantity",
+    theType: &FlaggedType {
+        element: IntType,
+        flags:   // ... elided ...
+    }
+}
+```
+
+Note how the type of `destinationEndpoint` includes some flags. This is a common model, and explains why we use `As*()` conversion methods rather than simple Go casts. Some of the wrapper types - such as `OptionalType` are important for our conversions, so we look out for those as we go.
+
+The function starts with checking prerequisites - the conditions under which this handler will trigger. Each bails out early if we find that the prerequisite is not satisfied.
+
+``` go
+    // Require both source and destination to not be bag items
+    if sourceEndpoint.IsBagItem() || destinationEndpoint.IsBagItem() {
+        return nil, nil
+    }
+
+    // Require both source and destination to be non-optional
+    if sourceEndpoint.IsOptional() || destinationEndpoint.IsOptional() {
+        return nil, nil
+    }
+```
+
+Neither of our endpoints is a bag item, and neither is optional, so our sample data passes these checks and the handler continues.
+
+Only primitive types can be simply copied across, so we check that both endpoints are the same primitive type.
+
+``` go
+    // Require source to be a primitive type
+    sourcePrimitive, sourceIsPrimitive := astmodel.AsPrimitiveType(sourceEndpoint.Type())
+    if !sourceIsPrimitive {
+        return nil, nil
+    }
+
+    // Require destination to be a primitive type
+    destinationPrimitive, destinationIsPrimitive := astmodel.AsPrimitiveType(destinationEndpoint.Type())
+    if !destinationIsPrimitive {
+        return nil, nil
+    }
+
+    // Require both properties to have the same primitive type
+    if !astmodel.TypeEquals(sourcePrimitive, destinationPrimitive) {
+        return nil, nil
+    }
+```
+
+When applied to `sourceEndPoint.Type()`, the `AsPrimitiveType()` conversion function returns the underlying `IntType`. It also knows how to unwrap our composite types, so when applied to `destinationEndPoint.Type()` it tunnels through the `FlaggedType` and also returns `IntType`. Since both types are the same, the last check passes as well.
+
+We finish by returning a function that creates the required code.
+
+``` go
+    return func(
+        reader dst.Expr, 
+        writer func(dst.Expr) []dst.Stmt, 
+        knownLocals *astmodel.KnownLocalsSet, 
+        generationContext *astmodel.CodeGenerationContext) []dst.Stmt {
+        return writer(reader)
+    }, nil
+}
+```
+
+* `reader` is an expression giving access to the original value
+* `writer` is a function takes an expression and writes that to the final destination
+* `knownLocals` is a set that contains all local variables declared so far (not used in this handler)
```

Comment:
This is a bit unclear as this puts a question in my mind regarding - local variables declared so far.. where? how?

Reply:
I've clarified.